### PR TITLE
feat(SDK): adds Artifact base class.

### PR DESCRIPTION
**Pull request by numerology**
**Description of your changes:**
Adds base `Artifact` Python class.

The base Python class mainly serves two purposes:

1. Help implementing the contract described in https://github.com/kubeflow/pipelines/blob/7a66414cf72ba5c746cac7fc6c8ecad67fc5e885/api/v2alpha1/pipeline_spec.proto#L397, so that user logic can access/manipulate MLMD properties easily during runtime.

2. Serve as a foundation for artifact ontology in the Python SDK

This implementation is similar to https://github.com/tensorflow/tfx/blob/master/tfx/types/artifact.py with two major differences:

1. KFP `Artifact` class is backed by `RuntimeArtifact` in its IR pb spec instead of Artifact pb message in MLMD.
2. The type schema is described as a YAML spec instead of pb spec. 

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

File: sdk/python/requirements.in [hunk 1]
```diff
@@ -21,3 +21,6 @@ requests_toolbelt>=0.8.0
 # CLI
 tabulate
 click
+
+# kfp.v2
+absl-py>=0.11.0,<1ß
```
Comment (chensun) [inline sdk/python/requirements.in:26]: remove `ß`?

File: sdk/python/kfp/v2/dsl/artifact.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+  """Property specified for an Artifact."""
+
+  # Mapping from Python enum to primitive type in the IR proto.
+  _ALLOWED_MLMD_TYPES = {
+      PropertyType.INT: pipeline_spec_pb2.PrimitiveType.INT,
+      PropertyType.DOUBLE: pipeline_spec_pb2.PrimitiveType.DOUBLE,
+      PropertyType.STRING: pipeline_spec_pb2.PrimitiveType.STRING,
+  }
+
+  def __init__(self, type, description: str = None):
```
Comment (chensun) [inline sdk/python/kfp/v2/dsl/artifact.py:47]: nit: `type: PropertyType` and `description: Optional[str] = None`

File: sdk/python/kfp/v2/dsl/artifact.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+  INT = 1
+  DOUBLE = 2
+  STRING = 3
+
+
+class Property(object):
+  """Property specified for an Artifact."""
+
+  # Mapping from Python enum to primitive type in the IR proto.
+  _ALLOWED_MLMD_TYPES = {
```
Comment (chensun) [inline sdk/python/kfp/v2/dsl/artifact.py:41]: nit: on the naming `_ALLOWED_MLMD_TYPES`, seems to me a bit misleading as I would think MLMD types are `Model`, `Dataset`, etc. 
Maybe this should be `_ALLOWED_PROPERTY_TYPES`? WDYT?

File: sdk/python/kfp/v2/dsl/artifact.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+      kind = PropertyType.DOUBLE
+    else:
+      raise TypeError('Got unknown type: %s' % dict_data['type'])
+
+    return Property(
+        type=kind,
+        description=dict_data['description']
+    )
+
+  def ir_type(self):
```
Comment (chensun) [inline sdk/python/kfp/v2/dsl/artifact.py:73]: nit: maybe `get_ir_type`? (more consistent with `get_type_name`)

File: sdk/python/kfp/v2/dsl/artifact.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+  _initialized = False
+
+  def __init__(self, instance_schema: Optional[str] = None):
+    """Constructs an instance of Artifact"""
+    if self.__class__ == Artifact:
+      if not instance_schema:
+        raise ValueError(
+            'The "instance_schema" argument must be passed to specify a '
+            'type for this Artifact.')
+      schema = yaml.safe_load(instance_schema)['properties']
```
Comment (chensun) [inline sdk/python/kfp/v2/dsl/artifact.py:125]: What if `instance_schema` is invalid? Should we check for `properties in ...`?

File: sdk/python/kfp/v2/dsl/artifact.py [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+      }
+    result_map = {
+        'title': title,
+        'type': 'object',
+        'properties': schema_map
+    }
+    return yaml.safe_dump(result_map)
+
+  @property
+  def type_schema(self):
```
Comment (chensun) [inline sdk/python/kfp/v2/dsl/artifact.py:164]: nit: `-> str`

File: sdk/python/kfp/v2/dsl/artifact.py [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+    return self._type_schema
+
+  def __repr__(self):
```
Comment (chensun) [inline sdk/python/kfp/v2/dsl/artifact.py:167]: nit: `-> str`

File: sdk/python/kfp/v2/dsl/artifact_test.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+
+_SERIALIZED_INSTANCE = """\
+{
+  "properties": {
+    "float1": {
+      "doubleValue": 1.11
+    },
+    "int1": {
+      "intValue": "1"
```
Comment (chensun) [inline sdk/python/kfp/v2/dsl/artifact_test.py:48]: re: `"1"`
Is quoting optional? Don't see quotes on `1.11`.

File: sdk/python/kfp/v2/dsl/artifact.py (hunk 3 above)
Comment (numerology) [reply sdk/python/kfp/v2/dsl/artifact.py:41]: Yep totally agree.

File: sdk/python/kfp/v2/dsl/artifact.py (hunk 2 above)
Comment (numerology) [reply sdk/python/kfp/v2/dsl/artifact.py:47]: Done

File: sdk/python/kfp/v2/dsl/artifact.py (hunk 4 above)
Comment (numerology) [reply sdk/python/kfp/v2/dsl/artifact.py:73]: Done

File: sdk/python/kfp/v2/dsl/artifact.py (hunk 5 above)
Comment (numerology) [reply sdk/python/kfp/v2/dsl/artifact.py:125]: Good point. Added a check

File: sdk/python/kfp/v2/dsl/artifact_test.py (hunk 8 above)
Comment (numerology) [reply sdk/python/kfp/v2/dsl/artifact_test.py:48]: No. It's interesting but I think double quotes are required here.

File: sdk/python/requirements.in (hunk 1 above)
Comment (numerology) [reply sdk/python/requirements.in:26]: Oh thanks for catching. Done

File: sdk/python/kfp/v2/dsl/artifact.py (hunk 6 above)
Comment (numerology) [reply sdk/python/kfp/v2/dsl/artifact.py:164]: Done

File: sdk/python/kfp/v2/dsl/artifact.py (hunk 7 above)
Comment (numerology) [reply sdk/python/kfp/v2/dsl/artifact.py:167]: Done

File: sdk/python/kfp/v2/dsl/artifact.py (hunk 3 above)
Comment (numerology) [reply sdk/python/kfp/v2/dsl/artifact.py:41]: Reworded. Resolving

**Review by numerology — COMMENTED** [review]
Thanks!

**Comment by chensun** [comment]
/lgtm

Thanks!

**Pull request merged**